### ISSUE TITLE
DINT propagation from high redshift

### ISSUE THREAD

**Issue by Mattver** [CRPropa/CRPropa3#174]
Hi,

I am having issues simulating a pure EM cascade from redshifts larger than about 0.63 (or distances of about 2400 Mpc).
Below this value, I am able to get a spectrum without an issue, but at higher values DINT returns empty histograms. Since the simulation time also goes down a lot beyond this value, I'm assuming there is some kind of cut-off? But I can't seem to find this in the code.

The simulation is set up to inject only photons at a certain redshift with a powerlaw spectrum. I'm using the EMCascade module to pass all photons to DINT immediately (I tried combined propagation, but altering the process method in a class inheriting from EMCascade doesn't seem to work). I did change the distance binning to have a maximum distance larger than the one I'm injecting the photons. Using the save method from EMCascade, I have verified that the histograms do contain photons before being passed to DINT.

Is there anything I am missing?

Thanks!

**Comment by rafaelab** [comment]
Hi!

I've tested it for z=1.2 and it works just fine, using the DintPropagation or EMCasc modules.
Are you sure you are passing the photon spectrum to Dint properly? Although if it works fo low redshifts, I don't see a reason why it would not work for higher redshifts.
Could you please provide a minimal working example that would allow me to reproduce the problem?

**Comment by Mattver** [comment]
Hi,

Here is the a small script that reproduces my code:


```python
from crpropa import *

def run(z, Emin, Emax, index, count):
    
    sim = ModuleList()

    sim.add(SimplePropagation())
    sim.add(Redshift())

    emcascade = EMCascade()
    dist = redshift2ComovingDistance(z)
    print dist/Mpc
    if dist > 1000*Mpc:
        emcascade.setDistanceBinning(1.001*dist, int(dist/Mpc))
    sim.add(emcascade)
    
    sim.add(MinimumEnergy(1*GeV))

    source = Source()
    source.add(SourceParticleType(22))
    source.add(SourceRedshift(z))
    source.add(SourcePosition(dist))
    source.add(SourcePowerLawSpectrum(Emin*GeV, Emax*GeV, index))

    sim.setShowProgress(True)
    sim.run(source, count, True)

    emcascade.save('emcascade_in.txt')
    emcascade.runCascade('emcascade_out.txt')

if __name__ == '__main__':

    run(0.65, 1e9, 1e11, -1, 10000)```

**Comment by rafaelab** [comment]
Thanks for the script!

If you vary z in your script you'll see that for z=0.63 you still get some non-zero bins at low energies. As z increases, the energy at which the bins start to have no events decrease. For z=0.65 all bins are 0, as you noted. If you take a look at the output, you'll see that for 1e7 eV the spectrum is more than 15 orders of magnitude larger than for 1e9 eV, suggesting a steep cutoff. The fact that you get zeros may be because:
1) The spectrum at higher energies is so low compared to the first bins (1e7 eV) that there isn't enough numerical precision to treat those values. This idea makes sense if you play with the values of z around 0.64.
2)  z=0.65 is quite high, meaning that most photons/electrons will interact before reaching Earth, transferring their energy to the low-energy region of the spectrum. 

I think the real issue you found is a combination of 1 and 2. If you simulate with an unrealistic B field of 1 T, you get non-zero bins in your spectrum.

In summary, I think this is not a real bug, it's just a limitation in the numerical precision of DINT. Does this make sense?

**Comment by Mattver** [comment]
Thanks a lot for the response!
This is what I was afraid of. I did some more tests and it seems that when you inject a lower energy spectrum, so less total energy injected, the cut off is at slightly lower z. So that seems to support the argument.

The fact that the simulation is 3x faster going from z=0.63 to 0.65 is then probably because of the minimum energy of the binning? Although naively I would have assumed that going beyond this 'maximum distance' the run time would stay roughly constant (at least at first).

Another issue I have is that if this is the explanation, I should get results up to higher redshift if I multiply all the counts in the emcascade_in.txt with 1e13. Instead it seems to stay zero for 0.65?
Something that also seems weird to me is that the when injecting low energy particles, the histogram seems to suddenly go from several counts to 0 (I guess first interaction makes the particle disappear completely?), whereas what I assume are cascaded particles go down to counts of 1e-10 and lower.

Finally, from what I could gather from skimming over the DINT code, it seems that it should be possible, at least in principle, to go to higher redshift. So I am a bit confused...

**Comment by Mattver** [comment]
So I tried looking into this a bit more, and leaving aside the fact that I don't understand why the runtime behaves like this etc, it might just be the correct result for this EBL model.
I was confused about the lack of photons around 100 GeV, because I was following arxiv.org/abs/1205.5755, where they state that the dominant contribution of GeV photons is from z~1-2. Also, some of the EBL models give rather low optical depths at those energies, even for high z (e.g. arxiv.org/abs/1212.1683 figure 9). However, the Stecker optical depth is higher (arxiv.org/abs/0510449 figure 8) at low energies. This might explain why the result is 0 for the simulation.
However, according to arxiv.org/abs/1005.0996 and arxiv.org/abs/1101.3498 the Stecker model is disfavored?

Testing the other available options (e.g. the low option), still doesn't give me events though. Although I don't know how this model compares to the default.

**Comment by avvliet** [comment]
The 'high', 'low' and 'primack' EBL options are described in the first CRPropa paper: astro-ph/0603675 Fig. 4.

**Comment by Mattver** [comment]
Hi,

Thanks for the reply! 
Looking at the Primack paper cited in the reference above (Fig. 3), I see that this model has an optical depth of around 1 for 100 GeV photons coming from z=1. Energies below 10 GeV are basically unattenuated. Instead, when looking at the simulation by CRPropa, the histograms are completely zero, even at energies below 1 GeV. So a more optimistic EBL model cannot be the solution.

As pointed out above, when increasing the magnetic field to unrealistic values, the resulting histograms are not empty (why?). In particular, the histograms get non-zero values starting at B=1e-5 T. However, the flux in the output is 48 orders of magnitude lower than the input flux, even at energies below 1 GeV. This is still far too much attenuation.

**Comment by rafaelab** [comment]
Hi,
I've done some tests and I still think the problem you've been finding relates to the numerical limitations of DINT. I'm not sure why it doesn't happend with strong B fields.
When you say you have scaled the produced photons (multiplied by 1e13), did you make sure this file was the one being read for propagation?

**Comment by Mattver** [comment]
Hi,

I just checked again and I'm indeed reading in the correct file (this time with a factor 1e10 amplification, also checked with a more reasonable factor of 1e5), see the script below. Of course if the problem is more than 48 orders of magnitude of suppression, this does not change anything. For the higher bins, this would then probably be the case.
Am I wrong to assume that if no interactions were to happen, the output should just contain the same number of photons as went in?
If I'm not wrong, then for sure if you inject photons in the lowest bins, they should (almost) keep their original values (and the same for photons that have been cascaded down)? I redid the same procedure for 1 GeV to 100 GeV for both the normal run and one where I multiplied the counts with 1e10, giving zero again. 
Otherwise, I am misunderstanding something basic.

[emcascade_in.txt](https://github.com/CRPropa/CRPropa3/files/1786364/emcascade_in.txt)


'''
from crpropa import *

def runagain(z):
    emcascade = EMCascade()
    dist = redshift2ComovingDistance(z)
    if dist > 1000*Mpc:
        emcascade.setDistanceBinning(1.001*dist, int(dist/Mpc))
    emcascade.load('emcascade_in.txt')
    emcascade.runCascade('emcascade_out.txt')

if __name__ == '__main__':

    #run(0.7, 1e9, 1e11, -1, 10000)
    runagain(0.7)
'''

**Comment by TobiasWinchen** [comment]
The problem is in the  RedshiftBinsDown method of DINT using steps far from end. Forcing fine-tuning steps solves the issue. I will commit a fix shortly.
